### DISCUSSION TITLE
Custom Actions 2.0 - Request for feedback

### DISCUSSION THREAD

**Discussion by kamilprusko** [focustimerhq/FocusTimer#714] [Show and tell]
I'll share what changes are coming to Custom Actions and I'd love some feedback.

# The current implementation (gnome-pomodoro 0.x)

Common issues with the current implementation:
1. Not intuitive, not descriptive enough
2. "Enabled/Disabled" made sense when the extension had a toggle
3. "Start" is ambiguous
4. "Resume" event firing when stopping a timer (if it was paused)
5. Not sure what values are inserted when using variables
6. No way to see what's happening - no built-in logging
7. Missing documentation / user guide

# Action list

The new actions list will have a link to a user guide (not written yet). The list is reorderable as it affects the order in which commands are executed, and there are quick toggles for the actions.
![actions-list](https://github.com/user-attachments/assets/4a021ed8-be9f-4a9e-b297-f0280223c0fa)

# Event trigger

More changes are in the action definition. Standard _Events_ like "Start" or "Pause" are fired only after clicking buttons or when using CLI (`gnome-pomodoro --start`) - there won't be phantom events like "Resume" fired when stopping the timer. The filter is hidden by default, and it may be finicky to use for some events. If you'd use filters, expect to debug it to see if it works...
![event-2](https://github.com/user-attachments/assets/835d1b32-1456-4a7c-9fdc-1fc24c1e13b7)

"Add Event"  popover:
![event-list](https://github.com/user-attachments/assets/b7380d31-9192-43fe-8aab-e9c45635b2dd)

Descriptions for the events:
- Start - Started the timer.
- Stop - Stopped the timer manually.
- Pause - The countdown has been manually paused. Not triggered when locking the screen or when suspending the system.
- Resume - The countdown has been manually resumed.
- Skip - Jumped to a next time-block before the countdown has finished.
- Rewind - Rewind action has been used. It adds a pause in the past.
- Reset - Manually cleared the session.
- Finished - The countdown has finished. If waiting for confirmation, the duration of the time-block still may be altered.
- Changed - Triggered on any change related to the countdown.
- Confirm Advancement - A manual confirmation is required to start next time-block.
- Advanced - Transitioned or skipped to a next time-block.
- State Changed - Transitioned to a next time-block or when a break gets relabelled.
- Rescheduled - Triggered on any change related to the session.
- Expired - Triggered when session is about to be reset due to inactivity.

# Condition trigger

Instead of defining a second action with counter triggers, you can now define a _Condition_. It has the benefit that it'll run the exit command more reliably and you don't have to guess what are the counter triggers.
![condition](https://github.com/user-attachments/assets/3a038e6b-cd57-44c9-92a4-270b36bb693b)

The condition can be nested up to one level, enough to mix AND and OR operands:
![condition-widget](https://github.com/user-attachments/assets/d698a054-00b8-402b-8f36-f7132c183a42)

# Commands and variables

There are more variables to insert into command line.
![variables](https://github.com/user-attachments/assets/154dd079-e239-42fd-923b-2d76673e1525)

...and you can choose a format:
![timestamp-variable](https://github.com/user-attachments/assets/32c97850-0f2d-4ba7-a6a4-6288a98024d4)
![state-variable](https://github.com/user-attachments/assets/b9ce8508-501d-41ee-b955-4f5e84283d56)

Notes:
- There is no longer variable listing the triggers.
- For `state`, `stopped` will be used instead of `null` and breaks may be represented by `break` (neither a short or long break).
- ISO8601 mentioned in "timestamp" popover is in UTC (not sure if it's a good default)
- There's a toggle "Pass Input Data" that will send JSON data to stdin, if you need the full context
- Commands are executed one by one in a queue. Each command has a timeout of 10s, unless you toggle off "Wait For Completion" option.

# Log

Finally, there's a built-in log. It can be accessed by the "Log" button on action list. Entries are not stored currently, data will be lost when closing the app. There's a limit to 200 entries.
![log](https://github.com/user-attachments/assets/f904c963-3b98-4f2a-bb2e-bac92192c83d)

If a command fails there will be a notification. If clicked, it opens the Log window too:
![notification](https://github.com/user-attachments/assets/f1ed076b-560d-49c7-8b2e-923f5917580e)

# Data migration

Unfortunately, I don't plan migrating actions from in the old app. It's extra development time and I doubt I will make it work for all cases, you would need to review it anyway. The old Pomodoro app and the new one can work side by side though.

# Try it out

I'm curious on your feedback.

Is it more intuitive or too overwhelming?
Are extra descriptions helpful or can be improved?
Is there an obstacle preventing from defining your custom actions?

**Comment by logrusx** [comment]
Hi,

I've been putting it off, but I finally went through your Request for Feedback. Honestly, nothing made sense to me. No that it doesn't make sense per se but because I don't use anything of it. I was just an extension user. What I needed it for was the timer, even the sound notifications I used very rarely to prompt me when it's time to start, when I was laying on the couch occasionally.

Having it in the top bar was very convenient, but when I found Hide Top Bar extension then it became less useful as I didn't see it. Now I don't use Gnome anymore I have no bars on the screen, only an OSD (pomodoro) status and OSD clock.

I don't even have frames around the windows as they take unnecessary space.

I hope that feedback is at least of some value.

Best Regards,
Georgi

**Reply by kamilprusko** [reply]
@logrusx thank you for your time. My hope is that improved builtin notifications should help in your use case.

**Reply by logrusx** [reply]
@kamilprusko , I have everything I need now, thanks to the script you provided in my issue.
